sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ReplaceCTERefWithRepartition.scala

```diff
@@ -41,19 +41,25 @@ object ReplaceCTERefWithRepartition extends Rule[LogicalPlan] {
       replaceWithRepartition(plan, mutable.HashMap.empty[Long, LogicalPlan])
   }
 
+  private def canSkipExtraRepartition(p: LogicalPlan): Boolean = p match {
+    case _: RepartitionOperation => true
+    case _: RebalancePartitions => true
+    case _ => false
+  }
+
   private def replaceWithRepartition(
       plan: LogicalPlan,
       cteMap: mutable.HashMap[Long, LogicalPlan]): LogicalPlan = plan match {
     case WithCTE(child, cteDefs) =>
       cteDefs.foreach { cteDef =>
         val inlined = replaceWithRepartition(cteDef.child, cteMap)
         val withRepartition =
-          if (inlined.isInstanceOf[RepartitionOperation] || cteDef.underSubquery) {
+          if (canSkipExtraRepartition(inlined) || cteDef.underSubquery) {
             // If the CTE definition plan itself is a repartition operation or if it hosts a merged
             // scalar subquery, we do not need to add an extra repartition shuffle.
             inlined
           } else {
-            Repartition(conf.numShufflePartitions, shuffle = true, inlined)
+            RepartitionByExpression(Seq.empty, inlined, None)
```

Comment:
what does this no-key shuffle do eventually?

Reply:
it uses RoundRobinPartitioning to do exchange which is same wih Repartition

Reply:
and the only difference is AQE can optimize it?

Reply:
yes, 

The eventually shuffle of Repartition is
`ShuffleExchangeExec(RoundRobinPartitioning, REPARTITION_BY_NUM)`

The eventually shuffle of RepartitionByExpression(Seq.empty) with none partition number is:
`ShuffleExchangeExec(RoundRobinPartitioning, REBALANCE_PARTITIONS_BY_NONE)`